### ISSUE TITLE
Mysterious problem compiling F* parser in Visual Studio 2013

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#325]
John Kennerley reports:
> ** Visual Studio
> I got clone https://github.com/FStarLang/FStar.git
> and tried to compile in vs2013, but got 3 compile errors.
> It was open on commit e8af9.
![image 1](https://cloud.githubusercontent.com/assets/5850655/9425427/fa23d284-48c4-11e5-9858-f72ac8d59caa.png)

**Comment by catalin-hritcu** [comment]
I cannot reproduce this on my (Windows virtual) machine, but I suspect that John's machine is missing dependency, probably fslex and fsyacc. Our instructions (https://github.com/FStarLang/FStar/blob/master/INSTALL.md#on-windows-78-using-visual-studio) only mention "VisualStudio 2013" and "Visual F# Tools (v3.0 or later)", but if I look at the extensions in Visual Studio then I also have NuGet, Visual F# 3.1.2, and Visual F# Power Tools installed:
<img width="960" alt="untitled" src="https://cloud.githubusercontent.com/assets/5850655/9425534/78259098-48c8-11e5-9c67-8ade9ca72bbb.png">
Could it be that some of these have to be installed manually, on top of "Visual F# Tools"?

**Comment by catalin-hritcu** [comment]
Also, if I look in my `FStar/bin` folder
```
hritcu@W7-VBox /cygdrive/c/Users/hritcu/Desktop/FStar
$ ls -al bin/
total 17652
drwxrwxr-x+ 1 hritcu None       0 Aug 22 21:22 .
drwxrwxr-x+ 1 hritcu None       0 Aug 15 01:38 ..
-rw-rwxr--+ 1 hritcu None      84 Aug 11 17:41 .gitignore
-rwxrwxr-x+ 1 hritcu None  753152 Aug 22 21:22 absyn.dll
-rwxrwxr-x+ 1 hritcu None  495104 Aug 22 21:22 absyn.pdb
-rwxrwxr-x+ 1 hritcu None  221184 Aug 22 21:22 backend.dll
-rwxrwxr-x+ 1 hritcu None  194048 Aug 22 21:22 backend.pdb
-rwxrwxr-x+ 1 hritcu None  190464 Aug 22 21:22 basic.dll
-rwxrwxr-x+ 1 hritcu None  226816 Aug 22 21:22 basic.pdb
-rwxrwxr-x+ 1 hritcu None  291328 Aug 22 21:22 extraction.dll
-rwxrwxr-x+ 1 hritcu None  251392 Aug 22 21:22 extraction.pdb
-rwxrwxr-x+ 1 hritcu None    8272 Aug 22 21:22 extraction.XML
-rwxrwxr-x+ 1 hritcu None   11264 Aug 22 21:22 format.dll
-rwxrwxr-x+ 1 hritcu None   19968 Aug 22 21:22 format.pdb
-rwxrwxr-x+ 1 hritcu None 1400472 Jun 15 14:02 FSharp.Core.dll
-rwxrwxr-x+ 1 hritcu None  618566 Aug 14  2014 FSharp.Core.xml
-rwxrwxr-x+ 1 hritcu None  359424 Mar 15 21:55 FSharp.PowerPack.Compatibility.dl                                                                                            l
-rwxrwxr-x+ 1 hritcu None  690176 Mar 15 21:55 FSharp.PowerPack.dll
-rwxrwxr-x+ 1 hritcu None   53760 Mar 15 21:55 FSharp.PPrint.dll
-rwxrwxr-x+ 1 hritcu None   71680 Mar 15 21:56 FsLexYacc.Runtime.dll
-rwxrwxr-x+ 1 hritcu None   15652 Mar 15 21:56 FsLexYacc.Runtime.xml
-rwxrwxr-x+ 1 hritcu None   29184 Aug 22 21:22 FStar.exe
-rwxrwxr-x+ 1 hritcu None     426 Aug 11 22:00 FStar.exe.config
-rwxrwxr-x+ 1 hritcu None   32256 Aug 22 21:22 FStar.pdb
-rwxrwxr-x+ 1 hritcu None      70 May 10 01:11 fstar-mono.sh
-rwxrwxr-x+ 1 hritcu None  188928 Aug 11 07:49 jsbackend.dll
-rwxrwxr-x+ 1 hritcu None  421200 Mar 15 21:55 msvcp100.dll
-rwxrwxr-x+ 1 hritcu None  770384 Mar 15 21:55 msvcr100.dll
-rwxrwxr-x+ 1 hritcu None  889344 Aug 22 21:22 parser.dll
-rwxrwxr-x+ 1 hritcu None  601600 Aug 22 21:22 parser.pdb
-rwxrwxr-x+ 1 hritcu None  621056 Aug 22 21:22 tc.dll
-rwxrwxr-x+ 1 hritcu None  544256 Aug 22 21:22 tc.pdb
-rwxrwxr-x+ 1 hritcu None   31232 Aug 22 21:22 tests.exe
-rwxrwxr-x+ 1 hritcu None     527 Aug 22 21:22 tests.exe.config
-rwxrwxr-x+ 1 hritcu None   44544 Aug 22 21:22 tests.pdb
-rwxrwxr-x+ 1 hritcu None  338432 Aug 22 21:22 tosmt.dll
-rwxrwxr-x+ 1 hritcu None  357888 Aug 22 21:22 tosmt.pdb
-rwxrwxr-x+ 1 hritcu None 7256576 Oct 27  2014 z3.exe
```
I can see DLLs that are not there in the github source (https://github.com/FStarLang/FStar/tree/master/bin) like `FsLexYacc.Runtime.dll`. The `Makefile` in the `VS` folder includes a call to ` .nuget/NuGet.exe restore FStar.sln` which seems to be bringing these extra files in. Is running `make` in the `VS` folder a prerequisite for building F* in VS that we should properly document?

**Comment by catalin-hritcu** [comment]
Reverted my `bin` folder to only the stuff that's on github, and even then building in VS worked well.  Rebuilding the parser used Nuget to bring in all the missing stuff:
```
------ Rebuild All started: Project: parser, Configuration: Debug Any CPU ------
	"C:\Users\hritcu\Desktop\FStar\VS\.nuget\NuGet.exe" install "C:\Users\hritcu\Desktop\FStar\src\parser\packages.config" -source ""  -NonInteractive -RequireConsent -solutionDir "C:\Users\hritcu\Desktop\FStar\VS\ "
	Restoring NuGet packages...
	To prevent NuGet from downloading packages during build, open the Visual Studio Options dialog, click on the Package Manager node and uncheck 'Allow NuGet to download missing packages'.
	All packages listed in packages.config are already installed.
	C:\Program Files\Microsoft SDKs\F#\3.1\Framework\v4.0\fsc.exe -o:obj\Debug\parser.dll -g --debug:full --noframework --define:TRACE --define:DEBUG --optimize+ -r:C:\Users\hritcu\Desktop\FStar\bin\absyn.dll -r:C:\Users\hritcu\Desktop\FStar\bin\basic.dll -r:C:\Users\hritcu\Desktop\FStar\bin\FSharp.PowerPack.Compatibility.dll -r:C:\Users\hritcu\Desktop\FStar\bin\FSharp.PowerPack.dll -r:C:\Users\hritcu\Desktop\FStar\VS\packages\FsLexYacc.Runtime.6.0.2\lib\net40\FsLexYacc.Runtime.dll -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\mscorlib.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Core.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Numerics.dll" -r:C:\Users\hritcu\Desktop\FStar\bin\tc.dll --target:library --warn:1 --warnaserror:76 --vserrors --validate-type-providers --LCID:1033 --utf8output --fullpaths --flaterrors --subsystemversion:6.00 --highentropyva+ --sqmsessionguid:0e138093-b1a3-48a2-8bcc-43b71436bcb1 --mlcompatibility "C:\Users\hritcu\AppData\Local\Temp\.NETFramework,Version=v4.5.AssemblyAttributes.fs" ast.fs parseutil.fs parse.fs lexhelp.fs lex.fs dsenv.fsi dsenv.fs desugar.fsi desugar.fs parseit.fsi parseit.fs driver.fs 
	parser -> C:\Users\hritcu\Desktop\FStar\bin\parser.dll
```
@John: could it be that you don't have nuget installed? (I have no clue whether it comes with VS or has to be brought in manually)

**Comment by catalin-hritcu** [comment]
This might be a duplicate of an old issue: #73. The error message at least seems very similar.

**Comment by catalin-hritcu** [comment]
Tried manually removing `src/parser/parse.fs` and `src/parser/lex.fs`, but rebuilding in Visual Studio properly regenerated these files by calling `fslex.exe` and `fsyacc.exe`:
```
------ Rebuild All started: Project: parser, Configuration: Debug Any CPU ------
	"C:\Users\hritcu\Desktop\FStar\VS\.nuget\NuGet.exe" install "C:\Users\hritcu\Desktop\FStar\src\parser\packages.config" -source ""  -NonInteractive -RequireConsent -solutionDir "C:\Users\hritcu\Desktop\FStar\VS\ "
	Restoring NuGet packages...
	To prevent NuGet from downloading packages during build, open the Visual Studio Options dialog, click on the Package Manager node and uncheck 'Allow NuGet to download missing packages'.
	All packages listed in packages.config are already installed.
	C:\Users\hritcu\Desktop\FStar\VS\packages\FsLexYacc.6.0.3\bin\fslex.exe -o lex.fs --unicode  lex.fsl 
	compiling to dfas (can take a while...)
	310 states
	writing output
	C:\Users\hritcu\Desktop\FStar\VS\packages\FsLexYacc.6.0.3\bin\fsyacc.exe -o parse.fs --module Microsoft.FStar.Parser.Parse  parse.fsy 
	building tables
	computing first function...time: 00:00:00.1251830
	building kernels...time: 00:00:00.0600265
	building kernel table...time: 00:00:00.0099660
	computing lookahead relations.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................time: 00:00:00.1862646
	building lookahead table...time: 00:00:00.0636631
	building action table...state 214: shift/reduce error on DOT
	state 251: shift/reduce error on SEMICOLON
	state 271: shift/reduce error on BAR
	state 276: shift/reduce error on BAR
	state 285: shift/reduce error on BAR
	state 377: shift/reduce error on LBRACE
	time: 00:00:00.2462843
	building goto table...time: 00:00:00.1297959
	returning tables.
	6 shift/reduce conflicts
	484 states
	133 nonterminals
	154 terminals
	290 productions
	#rows in action table: 484
	C:\Program Files\Microsoft SDKs\F#\3.1\Framework\v4.0\fsc.exe -o:obj\Debug\parser.dll -g --debug:full --noframework --define:TRACE --define:DEBUG --optimize+ -r:C:\Users\hritcu\Desktop\FStar\bin\absyn.dll -r:C:\Users\hritcu\Desktop\FStar\bin\basic.dll -r:C:\Users\hritcu\Desktop\FStar\bin\FSharp.PowerPack.Compatibility.dll -r:C:\Users\hritcu\Desktop\FStar\bin\FSharp.PowerPack.dll -r:C:\Users\hritcu\Desktop\FStar\VS\packages\FsLexYacc.Runtime.6.0.2\lib\net40\FsLexYacc.Runtime.dll -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\mscorlib.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Core.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Numerics.dll" -r:C:\Users\hritcu\Desktop\FStar\bin\tc.dll --target:library --warn:1 --warnaserror:76 --vserrors --validate-type-providers --LCID:1033 --utf8output --fullpaths --flaterrors --subsystemversion:6.00 --highentropyva+ --sqmsessionguid:0e138093-b1a3-48a2-8bcc-43b71436bcb1 --mlcompatibility "C:\Users\hritcu\AppData\Local\Temp\.NETFramework,Version=v4.5.AssemblyAttributes.fs" ast.fs parseutil.fs parse.fs lexhelp.fs lex.fs dsenv.fsi dsenv.fs desugar.fsi desugar.fs parseit.fsi parseit.fs driver.fs 
	parser -> C:\Users\hritcu\Desktop\FStar\bin\parser.dll
------ Rebuild All started: Project: backend, Configuration: Debug Any CPU ------
```
It used the `fslex.exe` and `fsyacc.exe` executables in `VS/packages/`:
```
hritcu@W7-VBox /cygdrive/c/Users/hritcu/Desktop/FStar
$ ls -l VS/packages/
total 1
drwxrwxr-x+ 1 hritcu None   0 Mar 15 21:56 FsLexYacc.6.0.3
drwxrwxr-x+ 1 hritcu None   0 Mar 15 21:56 FsLexYacc.Runtime.6.0.2
-rwxrwxr-x+ 1 hritcu None 132 Mar 15 22:40 repositories.config
```
Even when removing `VS/packages/` together with `src/parser/parse.fs` and `src/parser/lex.fs` rebuilding in Visual Studio properly brought back all files it needs to build the parser:
```
------ Rebuild All started: Project: parser, Configuration: Debug Any CPU ------
	"C:\Users\hritcu\Desktop\FStar\VS\.nuget\NuGet.exe" install "C:\Users\hritcu\Desktop\FStar\src\parser\packages.config" -source ""  -NonInteractive -RequireConsent -solutionDir "C:\Users\hritcu\Desktop\FStar\VS\ "
	Restoring NuGet packages...
	To prevent NuGet from downloading packages during build, open the Visual Studio Options dialog, click on the Package Manager node and uncheck 'Allow NuGet to download missing packages'.
	Installing 'FsLexYacc.Runtime 6.0.2'.
	Installing 'FsLexYacc 6.0.3'.
	Successfully installed 'FsLexYacc.Runtime 6.0.2'.
	Successfully installed 'FsLexYacc 6.0.3'.
	C:\Users\hritcu\Desktop\FStar\VS\packages\FsLexYacc.6.0.3\bin\fslex.exe -o lex.fs --unicode  lex.fsl 
	compiling to dfas (can take a while...)
	310 states
	writing output
	C:\Users\hritcu\Desktop\FStar\VS\packages\FsLexYacc.6.0.3\bin\fsyacc.exe -o parse.fs --module Microsoft.FStar.Parser.Parse  parse.fsy 
	building tables
	computing first function...time: 00:00:00.1180796
	building kernels...time: 00:00:00.0585816
	building kernel table...time: 00:00:00.0093891
	computing lookahead relations.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................time: 00:00:00.1745111
	building lookahead table...time: 00:00:00.0662730
	building action table...state 214: shift/reduce error on DOT
	state 251: shift/reduce error on SEMICOLON
	state 271: shift/reduce error on BAR
	state 276: shift/reduce error on BAR
	state 285: shift/reduce error on BAR
	state 377: shift/reduce error on LBRACE
	time: 00:00:00.2416047
	building goto table...time: 00:00:00.1364589
	returning tables.
	6 shift/reduce conflicts
	484 states
	133 nonterminals
	154 terminals
	290 productions
	#rows in action table: 484
	C:\Program Files\Microsoft SDKs\F#\3.1\Framework\v4.0\fsc.exe -o:obj\Debug\parser.dll -g --debug:full --noframework --define:TRACE --define:DEBUG --optimize+ -r:C:\Users\hritcu\Desktop\FStar\bin\absyn.dll -r:C:\Users\hritcu\Desktop\FStar\bin\basic.dll -r:C:\Users\hritcu\Desktop\FStar\bin\FSharp.PowerPack.Compatibility.dll -r:C:\Users\hritcu\Desktop\FStar\bin\FSharp.PowerPack.dll -r:C:\Users\hritcu\Desktop\FStar\VS\packages\FsLexYacc.Runtime.6.0.2\lib\net40\FsLexYacc.Runtime.dll -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\mscorlib.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Core.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.dll" -r:"C:\Program Files\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Numerics.dll" -r:C:\Users\hritcu\Desktop\FStar\bin\tc.dll --target:library --warn:1 --warnaserror:76 --vserrors --validate-type-providers --LCID:1033 --utf8output --fullpaths --flaterrors --subsystemversion:6.00 --highentropyva+ --sqmsessionguid:0e138093-b1a3-48a2-8bcc-43b71436bcb1 --mlcompatibility "C:\Users\hritcu\AppData\Local\Temp\.NETFramework,Version=v4.5.AssemblyAttributes.fs" ast.fs parseutil.fs parse.fs lexhelp.fs lex.fs dsenv.fsi dsenv.fs desugar.fsi desugar.fs parseit.fsi parseit.fs driver.fs 
	parser -> C:\Users\hritcu\Desktop\FStar\bin\parser.dll
```
So why does all this work just fine on my machine but fails on others?

**Comment by jkennerley** [comment]
Since I reported this issue several hours ago, I have walked rebooted my machine and generally done lots of other things.  I have come back to my machine and the solution now compiles.  Please close this issue. 

I have just opened the fstar solution again and recompiled and the solution compiles. 
Apologies for the erroneous issue.

I get this sometimes at work, where we have rather a large solution with many projects and extensive use of type providers. 
Our solution gives persistent files missing but several VS recompiles (ctrl-shift-B) gradually makes them go away.  
I was not expecting this on a small solution like fstar.sln. 
It is like the fsproj file knows about the file but the file is not on the disc. 
Anyway, it now compiles and I have another route forward.

**Comment by catalin-hritcu** [comment]
I have actually just been able to reproduce this with a fresh checkout and for me too a couple of reopens and recompiles made things eventually work. It would be good if we figure out what causes this general issue, so I'll leave it open even if your immediate problem seems solved now.

**Comment by jkennerley** [comment]
I have done a git pull and got  a new version of the repo.
It is on commit ddded.
After git reset --hard and git clean -xdf, I compiled and got a VS compile error as described above.
I tried 3 more times to compile, and the VS compile error did not go away. 
I tried closing the solution and re-opening the solution.
Then I tried the VS compile again and this time the VS compile worked with no errors.

git status shows this file is to be added on the next commit...
C:\FStar\src\extraction\bin\Debug\extraction.XML

So, the solution to the problem seems to be..
git pull
VS compile, 
if the solution gives the above compile errors then 
  close the solution
  open the solution and recompile in VS.

**Comment by markulf** [comment]
I could reproduce this as well. Here is my error message. 

```
------ Build started: Project: parser, Configuration: Debug Any CPU ------
	"C:\Users\markulf\Source\FStar\src\VS\.nuget\NuGet.exe" install "C:\Users\markulf\Source\FStar\src\parser\packages.config" -source ""  -NonInteractive -RequireConsent -solutionDir "C:\Users\markulf\Source\FStar\src\VS\ "
	Restoring NuGet packages...
	To prevent NuGet from downloading packages during build, open the Visual Studio Options dialog, click on the Package Manager node and uncheck 'Allow NuGet to download missing packages'.
	Installing 'FsLexYacc.Runtime 6.0.2'.
	Installing 'FsLexYacc 6.0.3'.
	Successfully installed 'FsLexYacc.Runtime 6.0.2'.
	Successfully installed 'FsLexYacc 6.0.3'.
C:\Program Files (x86)\MSBuild\12.0\bin\Microsoft.Common.CurrentVersion.targets(1697,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "FsLexYacc.Runtime". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors.
	C:\Program Files (x86)\Microsoft SDKs\F#\3.1\Framework\v4.0\fsc.exe -o:obj\Debug\parser.dll -g --debug:full --noframework --define:TRACE --define:DEBUG --optimize+ -r:C:\Users\markulf\Source\FStar\bin\absyn.dll -r:C:\Users\markulf\Source\FStar\bin\basic.dll -r:C:\Users\markulf\Source\FStar\bin\FSharp.PowerPack.Compatibility.dll -r:C:\Users\markulf\Source\FStar\bin\FSharp.PowerPack.dll -r:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\mscorlib.dll" -r:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Core.dll" -r:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.dll" -r:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Numerics.dll" -r:C:\Users\markulf\Source\FStar\bin\tc.dll --target:library --warn:1 --warnaserror:76 --vserrors --validate-type-providers --LCID:1033 --utf8output --fullpaths --flaterrors --subsystemversion:6.00 --highentropyva+ --sqmsessionguid:aa87d441-1b30-49b8-800a-4d18fca785ec --mlcompatibility "C:\Users\markulf\AppData\Local\Temp\.NETFramework,Version=v4.5.AssemblyAttributes.fs" ast.fs parseutil.fs parse.fs lexhelp.fs lex.fs dsenv.fsi dsenv.fs desugar.fsi desugar.fs parseit.fsi parseit.fs driver.fs 
FSC: error FS0225: Source file 'C:\Users\markulf\Source\FStar\src\parser\parse.fs' could not be found
FSC: error FS0225: Source file 'C:\Users\markulf\Source\FStar\src\parser\lex.fs' could not be found
Done building project "parser.fsproj" -- FAILED.
```

Closing the solution and opening it again worked. 
In the past I also did a "make -C src", to generate the missing `parse.fs` and `lex.fs` files, and then a rebuild in VS without closing the solution and this seemed to work as well.

**Comment by mlr-msft** [comment]
Should now be addressed in duplicate issue #430.
